platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+import jsonSchema from '@department-of-veterans-affairs/vets-json-schema/my-form-schema.json';
+
+const fromSchema = loadSchema(jsonSchema);
+
+export fromSchema;
+```
+
+**Single validate function for the page:**
+```js
+import fromSchema from '../schema-validation';
+
+const validate = values => {
+  const errors = {};
+  
+  const passwordErrorMessage = fromSchema('password'), // Path to schema
+    .minLength("This, uh, isn't very safe...") // Optional error message overrides
+    .pattern('Please follow the pattern, will you?') // Chaned message override functions
+    (values.password); // Data entered in the form
+  if (passwordErrorMessage) errors.password = passwordErrorMessage;
+}
+```
+
+**[Validate a single field:](https://formik.org/docs/guides/validation#field-level-validation)**
+```jsx
+import fromSchema from '../schema-validation';
+import { StringField } from 'formulate';
+
+const ThisPage = () => (
+  <>
+    <StringField
+      name="name.first"
+      validate={fromSchema('name.first').pattern("Please enter your _real_ name.")} />
+    <StringField
+      name="name.last"
+      validate={fromSchema('name.last')} />
+  </>
+)
+```
+
+Notice how in the first example, the function must be called with the value
+after all optional error message override functions, but in the second example,
+Formik calls the function with the value for the field itself.
+
+**Use the field value in the message:**
+```js
+const validateField = fromSchema('path')
+  .maxLength(value => `You entered ${value.length} characters, which is too many.`);
+```
+
+##### Caveat
+This field-level validation doesn't ensure that the form data submitted to the
+API matches the same shape as the JSON schema defines. Because a mismatch in the
+shape of the form data and corresponding schema has nothing to do with what the
+user entered, this is a developer mistake and should be caught by developers and
+not user-facing validation errors.
+
+To clarify, take the following example:
+**Schema**
+```json
+{
+  "type": "object",
+  "properties": {
+    "email": { "type": "string" },
+    "email-confirmation": { "type": "string" },
+  }
+}
+```
+
+**Data**
+```js
+{
+  email: {
+    original: "foo@bar.com",
+    confirmation: "foo@bar.com"
+  }
+}
+```
+Formulate will not deliberately catch mistakes like this, but it should be
+reasonably difficult to make these mistakes when using the `fromSchema`
+validation since the field's `name` property should likely match the schema path
+used in validation.
+
+Implementing such a mistake would look something like this using the
+`fromSchema` validation:
+```jsx
+<StringField
+  name="email-confirmation"
+  validation={fromSchema('email.confirmation')} />
+```
+
+**Discussion:** Is not ensuring the shape of the data matches the schema a
+reasonable boundary for the library to stop at?
+
+**Discussion:** What tools might Formulate give to developers to help catch
+these kinds of data shape mismatch mistakes?
```

Comment:
A linting rule, perhaps?

Reply:
Oh, that would be ideal. I'm not sure if it's feasible, though, since that schema may (probably will) come from an external package like `vets-json-schema`. It's worth exploring, though. :+1: